Cannot build rust when nested message contains repeated field

In this commit of my PR: https://github.com/pingcap/kvproto/pull/622/commits/e23c6bb183d4af51cceaf020957494d27b614294

And CI failed on building rust:

```
error[E0433]: failed to resolve: `super` in paths can only be used in start position
   --> /home/circleci/project/target/debug/build/kvproto-dc3c32c98dc79712/out/protos/wrapper_cdcpb.rs:385:65
    |
385 | #[inline] pub fn set_locks(&mut self, v: ::std::vec::Vec<event::super :: super :: kvrpcpb :: LockInfo>) { self.locks = v; }
    |                                                                 ^^^^^ `super` in paths can only be used in start position

error[E0433]: failed to resolve: `super` in paths can only be used in start position
   --> /home/circleci/project/target/debug/build/kvproto-dc3c32c98dc79712/out/protos/wrapper_cdcpb.rs:386:47
    |
386 | #[inline] pub fn get_locks(&self) -> &[event::super :: super :: kvrpcpb :: LockInfo] { &self.locks }
    |                                               ^^^^^ `super` in paths can only be used in start position

error[E0433]: failed to resolve: `super` in paths can only be used in start position
   --> /home/circleci/project/target/debug/build/kvproto-dc3c32c98dc79712/out/protos/wrapper_cdcpb.rs:387:70
    |
387 | #[inline] pub fn mut_locks(&mut self) -> &mut ::std::vec::Vec<event::super :: super :: kvrpcpb :: LockInfo> { &mut self.locks }
    |                                                                      ^^^^^ `super` in paths can only be used in start position

error[E0433]: failed to resolve: `super` in paths can only be used in start position
   --> /home/circleci/project/target/debug/build/kvproto-dc3c32c98dc79712/out/protos/wrapper_cdcpb.rs:388:66
    |
388 | #[inline] pub fn take_locks(&mut self) -> ::std::vec::Vec<event::super :: super :: kvrpcpb :: LockInfo> { ::std::mem::replace(&mut self.locks, ::std::vec::Vec::new()) }
    |                                                                  ^^^^^ `super` in paths can only be used in start position

error[E0412]: cannot find type `u64` in module `change_data_request`
   --> /home/circleci/project/target/debug/build/kvproto-dc3c32c98dc79712/out/protos/wrapper_cdcpb.rs:595:86
    |
595 | #[inline] pub fn set_txn_start_ts(&mut self, v: ::std::vec::Vec<change_data_request::u64>) { self.txn_start_ts = v; }
    |                                                                                      ^^^ not found in `change_data_request`
    |
help: possible candidates are found in other modules, you can import them into scope
    |
2   | use core::primitive::u64;
    |
2   | use std::primitive::u64;
    |

error[E0412]: cannot find type `u64` in module `change_data_request`
   --> /home/circleci/project/target/debug/build/kvproto-dc3c32c98dc79712/out/protos/wrapper_cdcpb.rs:596:68
    |
596 | #[inline] pub fn get_txn_start_ts(&self) -> &[change_data_request::u64] { &self.txn_start_ts }
    |                                                                    ^^^ not found in `change_data_request`
    |
help: possible candidates are found in other modules, you can import them into scope
    |
2   | use core::primitive::u64;
    |
2   | use std::primitive::u64;
    |

error[E0412]: cannot find type `u64` in module `change_data_request`
   --> /home/circleci/project/target/debug/build/kvproto-dc3c32c98dc79712/out/protos/wrapper_cdcpb.rs:597:91
    |
597 | #[inline] pub fn mut_txn_start_ts(&mut self) -> &mut ::std::vec::Vec<change_data_request::u64> { &mut self.txn_start_ts }
    |                                                                                           ^^^ not found in `change_data_request`
    |
help: possible candidates are found in other modules, you can import them into scope
    |
2   | use core::primitive::u64;
    |
2   | use std::primitive::u64;
    |

error[E0412]: cannot find type `u64` in module `change_data_request`
   --> /home/circleci/project/target/debug/build/kvproto-dc3c32c98dc79712/out/protos/wrapper_cdcpb.rs:598:87
    |
598 | #[inline] pub fn take_txn_start_ts(&mut self) -> ::std::vec::Vec<change_data_request::u64> { ::std::mem::replace(&mut self.txn_start_ts, ::std::vec::Vec::new()) }
    |                                                                                       ^^^ not found in `change_data_request`
    |
help: possible candidates are found in other modules, you can import them into scope
    |
2   | use core::primitive::u64;
    |
2   | use std::primitive::u64;
    |

error: aborting due to 8 previous errors
```

--- a/src/wrapper.rs
+++ b/src/wrapper.rs
@@ -456,7 +456,29 @@ impl FieldKind {
                     "::std::mem::replace(&mut self.{}, ::std::vec::Vec::new())",
                     result.name
                 ));
-                let unwrapped_type = unwrap_type(ty, "Vec").into_token_stream().to_string();
+                let mut unwrapped_type: &str =
+                    &unwrap_type(ty, "Vec").into_token_stream().to_string();
+
+                // The point of the prefix is to account for nesting of modules. However, it turns out
+                // that unwrapped_type may start with `super` so if we just smoosh the two together we
+                // get an invalid type. So the following small nightmare of code pops a suffix of
+                // prefix for every `super`, while there are both `super`s and segments of the prefix.
+                let mut segments: Vec<_> = prefix.split("::").collect();
+                while let Some(s) = segments.pop() {
+                    if s.is_empty() {
+                        continue;
+                    }
+                    if !unwrapped_type.starts_with("super ::") {
+                        segments.push(s);
+                        break;
+                    }
+                    unwrapped_type = &unwrapped_type[8..]
+                }
+                let mut prefix = segments.join("::");
+                if !prefix.is_empty() {
+                    prefix += "::"
+                }
+
                 result.ref_ty = RefType::Deref(format!("[{}{}]", prefix, unwrapped_type));
                 result.override_ty = Some(format!("::std::vec::Vec<{}{}>", prefix, unwrapped_type));
             }

